Framework/DataHandling/src/LoadInstrument.cpp

```diff
@@ -95,134 +98,162 @@ void LoadInstrument::init() {
  */
 void LoadInstrument::exec() {
   // Get the input workspace
-  m_workspace = getProperty("Workspace");
-  m_filename = getPropertyValue("Filename");
-  m_instName = getPropertyValue("InstrumentName");
+  boost::shared_ptr<API::MatrixWorkspace> ws = getProperty("Workspace");
+  std::string filename = getPropertyValue("Filename");
+  std::string instname = getPropertyValue("InstrumentName");
+  std::pair<std::string, std::string> loader_type;
 
-  // We will parse the XML using the InstrumentDefinitionParser
-  InstrumentDefinitionParser parser;
-
-  // If the XML is passed in via the InstrumentXML property, use that.
+  // If instrumentXML is not default (i.e. it has been defined), then use that
+  // Note: this is part of the IDF loader
   const Property *const InstrumentXML = getProperty("InstrumentXML");
   if (!InstrumentXML->isDefault()) {
     // We need the instrument name to be set as well because, for whatever
-    // reason,
-    //   this isn't pulled out of the XML.
-    if (m_instName.empty())
+    // reason, this isn't pulled out of the XML.
+    if (instname.empty())
       throw std::runtime_error("The InstrumentName property must be set when "
                                "using the InstrumentXML property.");
     // If the Filename property is not set, set it to the same as the instrument
     // name
-    if (m_filename.empty())
-      m_filename = m_instName;
-
-    // Initialize the parser. Avoid copying the xmltext out of the property
-    // here.
-    const PropertyWithValue<std::string> *xml =
-        dynamic_cast<const PropertyWithValue<std::string> *>(InstrumentXML);
-    if (xml) {
-      parser = InstrumentDefinitionParser(m_filename, m_instName, *xml);
-    } else {
-      throw std::invalid_argument("The instrument XML passed cannot be "
-                                  "casted to a standard string.");
-    }
-  }
-  // otherwise we need either Filename or InstrumentName to be set
-  else {
-    // Retrieve the filename from the properties
-    if (m_filename.empty()) {
+    if (filename.empty())
+      filename = instname;
+
+    loader_type.first = "idf";
```

Comment:
I'm not so happy with my approach here. I think that pair of strings is cumbersome, and the second values are not always used (the nexus for ex). I basically need to differentiate easily between 3 methods: xml, idf file and nxs. However, xml and idf file do a lot in common. I thought of assigning integers 1=xml, 2=idf, and 3=nxs and do the common stuff if the number is smaller than 3. But i find it makes the code less readable (you don't know right away what `loader_type = 1` means, `loader_type = xml` is much clearer). But the integer method is a lot more efficient. Any thoughts?

Reply:
I haven't read the code, but it sounds like you want an `enum` (likely new style). If you want to perform some later clean-up of this kind, you could defer to another PR? Up to you.